network/noise/src/noise_fuzzing.rs

```diff
@@ -0,0 +1,293 @@
+    }
+
+    fn next_u64(&mut self) -> u64 {
+        self.0 += 1;
+        self.0
+    }
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        impls::fill_bytes_via_next(self, dest)
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand_core::Error> {
+        Ok(self.fill_bytes(dest))
+    }
+}
+impl CryptoRng for CountingRng {}
+impl Random for CountingRng {}
+
+// TestResolver is needed to have a deterministic RNG in keypair generation
+struct TestResolver {}
+impl CryptoResolver for TestResolver {
+    fn resolve_rng(&self) -> Option<Box<dyn Random>> {
+        let rng = CountingRng(0);
+        Some(Box::new(rng))
+    }
+
+    fn resolve_dh(&self, choice: &DHChoice) -> Option<Box<dyn Dh>> {
+        DefaultResolver.resolve_dh(choice)
+    }
+
+    fn resolve_hash(&self, choice: &HashChoice) -> Option<Box<dyn Hash>> {
+        DefaultResolver.resolve_hash(choice)
+    }
+
+    fn resolve_cipher(&self, choice: &CipherChoice) -> Option<Box<dyn Cipher>> {
+        DefaultResolver.resolve_cipher(choice)
+    }
+}
+
+// ExposingSocket is needed to retrieve the noise messages peers will write on the socket
+struct ExposingSocket {
+    pub inner: MemorySocket,
+    pub written: Vec<u8>,
+}
+impl ExposingSocket {
+    fn new(memory_socket: MemorySocket) -> Self {
+        Self {
+            inner: memory_socket,
+            written: Vec::new(),
+        }
+    }
+    fn new_pair() -> (Self, Self) {
+        let (dialer_socket, listener_socket) = MemorySocket::new_pair();
+        (
+            ExposingSocket::new(dialer_socket),
+            ExposingSocket::new(listener_socket),
+        )
+    }
+}
+impl AsyncWrite for ExposingSocket {
+    //    use AsyncWrite;
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        context: &mut Context,
+        buf: &[u8],
+    ) -> Poll<futuresResult<usize>> {
+        // write to us first
+        self.written.extend_from_slice(buf);
+        //
+        Pin::new(&mut self.inner).poll_write(context, buf)
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<futuresResult<()>> {
+        Pin::new(&mut self.inner).poll_flush(context)
+    }
+
+    /// Attempt to close the channel. Cannot Fail.
+    fn poll_close(mut self: Pin<&mut Self>, context: &mut Context) -> Poll<futuresResult<()>> {
+        Pin::new(&mut self.inner).poll_close(context)
+    }
+}
+impl AsyncRead for ExposingSocket {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        context: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<futuresResult<usize>> {
+        Pin::new(&mut self.inner).poll_read(context, buf)
+    }
+}
+
+// take_socket is implemented on parent structure to be able to retrieve our ExposingSocket
+impl NoiseSocket<ExposingSocket> {
+    fn take_socket(self) -> ExposingSocket {
+        self.socket
+    }
+}
+
+//
+// Actual code to generate corpus
+//
+
+// let's cache the deterministic keypair
+use once_cell::sync::Lazy;
```

Comment:
move this import to the top

Reply:
isn't it clearer here? Since it is only used in one place